internal/component/prometheus/remote/queue/network/loop.go

```diff
@@ -0,0 +1,374 @@
+	result := l.send(ctx, attempts)
+	duration := time.Since(start)
+	l.statsFunc(types.NetworkStats{
+		SendDuration: duration,
+	})
+	if result.successful {
+		l.sendingCleanup()
+		return
+	}
+	if !result.recoverableError {
+		l.sendingCleanup()
+		return
+	}
+	attempts++
+	if attempts > int(l.cfg.MaxRetryBackoffAttempts) && l.cfg.MaxRetryBackoffAttempts > 0 {
+		level.Debug(l.log).Log("msg", "max retry attempts reached", "attempts", attempts)
+		l.sendingCleanup()
+		return
+	}
+	// This helps us short circuit the loop if we are stopping.
+	if l.stopCalled.Load() {
+		return
+	}
+	// Sleep between attempts.
+	time.Sleep(result.retryAfter)
+	goto attempt
+}
+
+type sendResult struct {
+	err              error
+	successful       bool
+	recoverableError bool
+	retryAfter       time.Duration
+	statusCode       int
+	networkError     bool
+}
+
+func (l *loop) sendingCleanup() {
+	types.PutTimeSeriesSliceIntoPool(l.series)
+	l.sendBuffer = l.sendBuffer[:0]
+	l.series = make([]*types.TimeSeriesBinary, 0, l.batchCount)
+	l.lastSend = time.Now()
+}
+
+// send is the main work loop of the loop.
+func (l *loop) send(ctx context.Context, retryCount int) sendResult {
+	result := sendResult{}
+	defer func() {
+		recordStats(l.series, l.isMeta, l.statsFunc, result, len(l.sendBuffer))
+	}()
+	var err error
+	// Check to see if this is a retry and we can reuse the buffer.
+	// I wonder if we should do this, its possible we are sending things that have exceeded the TTL.
+	if len(l.sendBuffer) == 0 {
+		var data []byte
+		var wrErr error
+		if l.isMeta {
+			data, wrErr = createWriteRequestMetadata(l.req, l.series, l.buf)
+		} else {
+			data, wrErr = createWriteRequest(l.req, l.series, l.externalLabels, l.buf)
+		}
+		if wrErr != nil {
+			result.err = wrErr
+			result.recoverableError = false
+			return result
+		}
+		l.sendBuffer = snappy.Encode(l.sendBuffer, data)
+	}
+
+	httpReq, err := http.NewRequest("POST", l.cfg.URL, bytes.NewReader(l.sendBuffer))
+	if err != nil {
+		result.err = err
+		result.recoverableError = true
+		result.networkError = true
+		return result
+	}
+	httpReq.Header.Add("Content-Encoding", "snappy")
+	httpReq.Header.Set("Content-Type", "application/x-protobuf")
+	httpReq.Header.Set("User-Agent", l.cfg.UserAgent)
+	httpReq.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
+	httpReq.SetBasicAuth(l.cfg.Username, l.cfg.Password)
+
+	if retryCount > 0 {
+		httpReq.Header.Set("Retry-Attempt", strconv.Itoa(retryCount))
+	}
+	ctx, cncl := context.WithTimeout(ctx, l.cfg.Timeout)
+	defer cncl()
+	resp, err := l.client.Do(httpReq.WithContext(ctx))
+	// Network errors are recoverable.
+	if err != nil {
+		result.err = err
+		result.networkError = true
+		result.recoverableError = true
+		result.retryAfter = l.cfg.RetryBackoff
+		return result
+	}
+	result.statusCode = resp.StatusCode
+	defer resp.Body.Close()
+	// 500 errors are considered recoverable.
+	if resp.StatusCode/100 == 5 || resp.StatusCode == http.StatusTooManyRequests {
```

Comment:
In vanilla remote_write we had a setting to "retry on 429" - do you think we'd need it here?

Reply:
I think that is something we can look into, currently we default to true in remote_write vanilla. If we need to we can add a config to turn it off.